### ISSUE TITLE
Handle OSS licenses and attribution

### ISSUE THREAD

**Issue by guydc** [envoyproxy/gateway#2917]
*Description*:
Envoy Gateway should adhere to [CNCF guidelines for dependency licenses](https://github.com/cncf/foundation/blob/main/allowed-third-party-license-policy.md). As an Apache-2.0 project, EG may have [licensing implications](https://fossa.com/blog/open-source-software-licenses-101-gpl-v3/) from using copyleft dependencies that are licensed under GPLv3. 

Some projects enforce these guidelines using tools like [go-licenses](https://github.com/google/go-licenses), e.g. [cilium](https://github.com/cilium/cilium/tree/main/tools/licensecheck). Automated validation of of approved licenses should be a part of the PR approval flow. 

Additionally, the CNCF guidelines for [OSS attribution](https://github.com/cncf/foundation/blob/main/recommendations-for-attribution.md) should be followed, by generating and maintaining the appropriate documentation.

**Comment by arkodg** [comment]
is this something that can be built into the openssf tooling ?

**Comment by guydc** [comment]
Not sure. There's this issue: https://github.com/ossf/scorecard/issues/3840, but if focuses on the license in use by the project, not its dependencies. There's this issue: https://github.com/ossf/scorecard/issues/2531 that deals with dependencies. 

IMHO, this should be a reusable CNCF util, as it's meant to enforce CNCF-specific requirements.

**Comment by guydc** [comment]
@shahar-h - there are two options that we discussed in this context:
- Making permissive dependency license checks part of the OSSF project
- Trying to build a common CNCF util

**Comment by shahar-h** [comment]
@guydc [FOSSA](https://github.com/fossas/fossa-cli) is commonly used by CNCF projects: https://github.com/cncf/foundation/issues/109

**Comment by arkodg** [comment]
+1 for FOSSA, will reach out to CNCF to get a key

**Comment by shahar-h** [comment]
@arkodg @guydc 
I tested `FOSSA` on my EG fork repo and have some questions/insights:
1. `FOSSA` free license supports max 5 dependency depth levels, where business/enterprise license have no limit. `EG` dependency depth level is 9. Can you tell which license CNCF work with?
1. In order to trigger `FOSSA` scan also on PRs api key must be hardcoded in the workflow, since GitHub prevents PRs from forks from accessing secrets by default(See related [discussion](https://github.com/distribution/distribution/pull/3396#discussion_r617361230)).
We have 2 options here:
    1. Trigger `FOSSA` scan only on push to `main` events, which will allow us to read the token from secret.
    1. Trigger `FOSSA` scan also on PRs and expose the api token.   
WDYT?
1. I used [FOSSA official GH action ](https://github.com/marketplace/actions/official-fossa-action), which can be used in 2 modes:
    1. Non-blocking - scans dependencies, uploads report to server but doesn't wait for completion.
    1. Blocking -  scans dependencies, uploads report to server and polling the server for status until it's completed.    
I guess that we want to work in blocking mode in order to fail the build on any new issue. 
1. The scan found some issues, which all look false positives. I can share my analysis if needed.  Because of this I think we  should start with non-blocking mode, triage all issues and then switch to blocking mode.

**Comment by shahar-h** [comment]
Update: FOSSA scan was done against some default license policy which is not compliant with CNCF policy.
Since I can't change the policy in the free version I go over scan results and found 2 license violations:
* github.com/hashicorp/errwrap
* github.com/hashicorp/go-multierror

Both declare `MPL-2.0` license which is not included in [CNCF approved licenses list](https://github.com/cncf/foundation/blob/main/allowed-third-party-license-policy.md#approved-licenses-for-allowlist).

Both dependencies are transitive and used by `helm.sh/helm/v3`, which is used by `egctl`.
Anyway there is an open [PR](https://github.com/helm/helm/pull/12818) that removes both dependencies and targeted to `3.15.0` release.

BTW [osv-scanner](https://google.github.io/osv-scanner/) which is used by [OpenSSF](https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#vulnerabilities) for vulnerability scanning also supports [license scanning](https://google.github.io/osv-scanner/experimental/license-scanning/). I tested it and it found the same violations with some few false positives. The problem that it's still experimental. We can consider using it in the future.

**Comment by arkodg** [comment]
thanks for digging into this @shahar-h !
+1 on non blocking scans until we fix all

I'm still waiting to hear back from CNCF, will reach out again

**Comment by shahar-h** [comment]
Thanks, and what about the second point(api token)?

**Comment by arkodg** [comment]
> Thanks, and what about the second point(api token)?

yeah read the context from the distribution PR, CNCF + FOSSA is a +1 with that approach so we can just follow it, also include them in the PR, to make sure we're going the right thing

**Comment by RobertKielty** [comment]
@Xunzhuo logged a CNCF Service Desk request.

I invited (most) all of the maintainers to join the CNCF FOSSA Organizaion.

FOSSA Invites _are only valid for 48 hours_. https://docs.fossa.com/docs/inviting-users

@Xunzhuo please spread the word and ask the invitees to accept their invitations.

I need _just one maintainer_ to accept the invite who can then add repos following these instructions:

https://docs.fossa.com/docs/quick-import

Any issues reach out to me on CNCF Slack.

**Comment by shahar-h** [comment]
Regarding `osv-scanner` option, there are a few gaps:
* There are some missing packages and versions in [deps.dev](https://deps.dev/) metadata. Opened a few [issues](https://github.com/google/deps.dev/issues/created_by/shahar-h).
* We are waiting for helm.sh/helm/v3 v3.15.0 release in order to resolve hashicorp packages violations: https://github.com/helm/helm/pull/12818.
* github.com/grafana/tempo which is used in e2e tests declares a disallowed license. We can either replace/remove it or wait for https://github.com/google/osv-scanner/pull/944 that will add support for package exclusion.
* https://github.com/google/osv-scanner/pull/944 also adds a capability of license overriding which may be useful in case of missing packages/versions in `deps.dev` or dual licenses.

**Comment by arkodg** [comment]
thanks for driving this work @shahar-h !

For other following this issue wanted to highlight why Envoy Gateway is trying to adopt `osv-scanner` over `FOSSA`
* FOSSA is a SaaS tool that has its own notion of user management, for a open source project, which is a collection of loosely coupled maintainers, managing users in multiple places (apart from GH) adds maintenance burden
* reviewing issues/violations must be done in the FOSSA platform instead of within GH where user roles are already assigned to view the right apps
* the FOSSA token is tied to a email/user, Envoy Gateway does not have any bot emails, and after chatting with @phlax even Envoy Proxy is trying to move away from managing bot emails, this adds more maintenance once that user leaves the project

having this said, if we cannot achieve the goal of license scanning via `osv-scanner` , we may need to move to FOSSA

**Comment by shahar-h** [comment]
A more detailed context about `osv-scanner` gaps:
Current scan results:
<img width="819" alt="image" src="https://github.com/envoyproxy/gateway/assets/16555289/91db28b4-d2cd-4c51-bf51-ac297075c498">
* All `UNKNOWN` license violation are a result of missing package/version in [deps.dev](https://deps.dev/) service. I opened an issue for each one in `deps.dev` repo:
  * https://github.com/google/deps.dev/issues/83
  * https://github.com/google/deps.dev/issues/84
  * https://github.com/google/deps.dev/issues/85
  * https://github.com/google/deps.dev/issues/86
  * https://github.com/google/deps.dev/issues/87
We should be able to manually override the license in these cases once https://github.com/google/osv-scanner/pull/944 is merged.
* Regarding `CC-BY-SA-4.0` license violation for `github.com/docker/go-metrics` and `github.com/opencontainers/go-digest`: both have a dual license: the code is licensed under the `Apache 2.0` license and the docs under `CC-BY-SA-4.0` license. See:
https://github.com/docker/go-metrics?tab=readme-ov-file#copyright-and-license
https://github.com/opencontainers/go-digest?tab=readme-ov-file#copyright-and-license
`osv-scanner` doesn't make the difference between license of choice and multiple licenses so it fails for every disallowed license in the list. See explanation [here](https://github.com/google/osv-scanner/issues/949#issuecomment-2087956708).  Anyway we should be able to choose the desired license in these cases once https://github.com/google/osv-scanner/pull/944 is merged. 
* `github.com/hashicorp/errwrap` and `github.com/hashicorp/go-multierror` are licensed under `MPL-2.0` license, which is disallowed. Both will be removed once https://github.com/helm/helm/pull/12818 is merged and `helm` `3.15.0` is released. We can temporarily ignore helm package once https://github.com/google/osv-scanner/pull/944 is merged.
* `github.com/grafana/tempo` is licensed under `AGPL-3.0` license, which is disallowed, but since it's only used in e2e tests we can safely ignore it once https://github.com/google/osv-scanner/pull/944 is merged. We can also consider replacing it with some other package so we don't use it by accident in the future in released code.

**Comment by guydc** [comment]
This was discussed in the community meeting. For the time being, we can proceed with:
- Implement an async, non-blocking osv scan workflow action
- Create another issue to track all of the aforementioned limitations and their status. When all issues are resolved, we can turn on enforcement.

**Comment by shahar-h** [comment]
There is an additional issue: osv license scan can only be run together with vulnerability scan. Since we want to run vulnerability scan recursively(to be aligned with `ossf-scorecard`) but license scan only on root folder(I guess we don't care about dependencies licenses in tools and site folders) we either need an option to run license scan independently or ask support for a dedicated recursive flag for each scan. See related [issue](https://github.com/google/osv-scanner/issues/965).
